components/layout_thread/lib.rs

```diff
@@ -928,47 +932,56 @@ impl LayoutThread {
 
             if flow::base(layout_root).restyle_damage.contains(REPAINT) ||
                     rw_data.display_list.is_none() {
-                let mut root_stacking_context = StackingContext::new(StackingContextId::new(0),
-                                                                     StackingContextType::Real,
-                                                                     &Rect::zero(),
-                                                                     &Rect::zero(),
-                                                                     0,
-                                                                     filter::T::new(Vec::new()),
-                                                                     mix_blend_mode::T::normal,
-                                                                     Matrix4D::identity(),
-                                                                     Matrix4D::identity(),
-                                                                     true,
-                                                                     false,
-                                                                     None);
-
-                let display_list_entries =
-                    sequential::build_display_list_for_subtree(layout_root,
-                                                               &mut root_stacking_context,
-                                                               shared_layout_context);
-
-                debug!("Done building display list.");
-
-                let root_background_color = get_root_flow_background_color(layout_root);
-                let root_size = {
-                    let root_flow = flow::base(layout_root);
-                    if rw_data.stylist.viewport_constraints().is_some() {
-                        root_flow.position.size.to_physical(root_flow.writing_mode)
-                    } else {
-                        root_flow.overflow.scroll.size
+                let display_list_needed = query_type.map(reflow_query_type_needs_display_list)
+                                                    .unwrap_or(false);
+                match (data.goal, display_list_needed) {
+                    (ReflowGoal::ForDisplay, _) | (ReflowGoal::ForScriptQuery, true) => {
```

Comment:
Am I right in guessing that the `query_type` is `Some` if and only if `data.goal == ReflowGoal::ForScriptQuery`?

If so, either add an assertion for that `assert!(query_type.is_some() == data.goal == ReflowGoal::ForScriptQuery)`, or put the query type in the script reflow goal. The latter's more rustic, but might be complicated enough to save for another PR.